[exporter/syslog] send syslog messages in batches

andrzej-stencel opened the pull request:
**Description:**

This changes the behavior of the Syslog exporter to send each batch of Syslog messages in a single request (with messages separated by newlines), instead of sending each message in a separate request and closing the connection after each message.

The batching only happens when using TCP. For UDP, each syslog message is still sent in a separate request, as defined by [the spec](https://datatracker.ietf.org/doc/html/rfc5426#section-3.1).

This also significantly refactors (and hopefully simplifies) the exporter's code, extracting the code that formats the syslog messages from the `sender` type into separate `formatter` types. Hopefully this will make the development of this component easier.

**Link to tracking Issue:**

- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21244

**Testing:**

The unit tests have been updated to reflect the refactored codebase. The integration tests introduced in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/27464 are unchanged, as the format of the output messages hasn't changed.

**Documentation:**

No documentation updates.

andrzej-stencel committed:
[exporter/syslog] send syslog messages in batches

This changes the behavior of the Syslog exporter
to send each batch of syslog messages in a single request
(with messages separated by newlines),
instead of sending each message in a separate request
and closing the connection after each message.

This also significantly refactors (and hopefully simplifies) the exporter's code,
extracting the code that formats the Syslog messages
from the `sender` type into separate `formatter` types.
Hopefully this will make the development of this component easier.

rnishtala-sumo commented on exporter/syslogexporter/exporter.go:
```diff
@@ -69,78 +68,30 @@ func newLogsExporter(
+			for k := 0; k < scopeLogs.LogRecords().Len(); k++ {
+				logRecord := scopeLogs.LogRecords().At(k)
+				formatted := se.formatter.format(logRecord)
+				payload.WriteString(formatted)
```
should we have a configurable limit on the payload size, now that we're batching log messages?

astencel-sumo replied:
Good point that it should be possible to limit the payload size. Users should be able to use the collector's built-in batching for this. Currently this functionality is provided the [Batch processor](https://github.com/open-telemetry/opentelemetry-collector/tree/main/processor/batchprocessor). In future, batching hopefully will be a part of the [exporter helper](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/exporterhelper).

rnishtala-sumo replied:
oh right, I forgot that this is an exporter and not the receiver.

andrzej-stencel committed:
Merge branch 'main' into syslog-exporter-performance

andrzej-stencel committed:
Merge branch 'main' into syslog-exporter-performance

andrzej-stencel committed:
Merge branch 'main' into syslog-exporter-performance

andrzej-stencel committed:
Merge branch 'main' into syslog-exporter-performance

andrzej-stencel committed:
Merge branch 'main' into syslog-exporter-performance

andrzej-stencel committed:
fix: do not batch messages when sending over UDP

According to https://datatracker.ietf.org/doc/html/rfc5426#section-3.1,
Messages sent over UDP must be sent separately.

> Each syslog UDP datagram MUST contain only one syslog message

astencel-sumo commented:
I have changed the implementation to only batch messages when using TCP. As correctly pointed out by @kkujawa-sumo (thank you!), when sending over UDP, each message must be sent in a separate request: https://datatracker.ietf.org/doc/html/rfc5426#section-3.1

> Each syslog UDP datagram MUST contain only one syslog message

andrzej-stencel committed:
Merge branch 'main' into syslog-exporter-performance

andrzej-stencel committed:
Merge branch 'main' into syslog-exporter-performance

rnishtala-sumo commented on exporter/syslogexporter/formatter.go:
```diff
@@ -0,0 +1,29 @@
+}
+
+type formatter interface {
+	format(plog.LogRecord) string
```
should the formatter interface have a validation function that returns an error in the following scenarios

- a required field is empty
- The value for a field is not in the right format. For example:
   ```
   The PRI part MUST have three, four, or five characters and will be
   bound with angle brackets as the first and last characters.  The PRI
   part starts with a leading "<" ('less-than' character, %d60),
   followed by a number, which is followed by a ">"
```

kkujawa-sumo replied:
> should the formatter interface have a validation function that returns an error in the following scenarios
> 
> * a required field is empty
> * The value for a field is not in the right format. For example:
>   ```
>   The PRI part MUST have three, four, or five characters and will be
>   bound with angle brackets as the first and last characters.  The PRI
>   part starts with a leading "<" ('less-than' character, %d60),
>   followed by a number, which is followed by a ">"
>   ```

`format` function will always add the PRI part e.g. [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/27799/files#diff-fd44958375c0625e37038008e6afcd03ad386b6e60db4fb1d4d1fdccb1c4a27dR30)

for example [time.Format ](https://pkg.go.dev/time#Time.Format) doesn't return any error

rnishtala-sumo replied:
makes sense, the formatter ensures a valid format for PRI and other fields, I wonder what kind of validation logic exists in the syslog receiver. The exporter probably doesn't need to do any field validation (apart from populating default values).

astencel-sumo replied:
I initially implemented the `formatter`'s `format` function to return `(string, error)` but then it occurred that the implementation doesn't ever return an error. As there is no erroneous input, there's no need for a validation function.

andrzej-stencel committed:
Merge branch 'main' into syslog-exporter-performance

astencel-sumo commented on .chloggen/syslog-exporter-single-request.yaml:
```diff
@@ -0,0 +1,27 @@
+component: exporter/syslog
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: send syslog messages in batches
```
Good point, updated.

astencel-sumo commented on .chloggen/syslog-exporter-single-request.yaml:
```diff
@@ -0,0 +1,27 @@
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: This changes the behavior of the Syslog exporter to send each batch of Syslog messages in a single request (with messages separated by newlines), instead of sending each message in a separate request and closing the connection after each message.
```
Updated.

astencel-sumo commented on exporter/syslogexporter/rfc5424_formatter_test.go:
```diff
@@ -0,0 +1,94 @@
+	"go.opentelemetry.io/collector/pdata/plog"
+)
+
+func TestRFC5424Formatter(t *testing.T) {
```
The exporter never fails on formatting the message. It will use default values if anything is missing.

andrzej-stencel committed:
Merge branch 'main' into syslog-exporter-performance

andrzej-stencel committed:
Merge branch 'main' into syslog-exporter-performance

andrzej-stencel committed:
Merge branch 'main' into syslog-exporter-performance

andrzej-stencel committed:
Merge branch 'main' into syslog-exporter-performance

astencel-sumo commented:
@songy23 anything missing from marking it as `ready to merge`? Not sure if this is the way to signal to maintainers that this can be merged.

songy23 commented:
Added the label, cc @open-telemetry/collector-contrib-maintainer please merge 🙏
